### ISSUE TITLE
Decorator to facilitate config changes in tests

### ISSUE THREAD

**Issue by wordofglass** [beetbox/beets#1950]
While modifying some of `fetchart`'s tests I came across the problem that test methods of one class need to apply different configurations each. To make this work, `fetchart` currently reads parts of the config not in `__init__` but when acutally searching for artwork. This is because the `UseThePlugin` subclass of `TestCase` that all tests use in order not to duplicate the plugin loading code otherwise would initialize the plugin (and read the config) before it can be changed in the method

It would probably be nicer to have all the config reading and validation in `__init__`, though. I don't know how much this affects other parts of beets, but I'd assume the problem to be global. One solution I came up with is the following decorator (which I did not test, maybe it doesn't even work!). The only catch might be that it makes use of `unittest.TestCase._testMethodName`, which is undocumented:

The decorator
```python
    def with_config(self, func, config, before_setup=True):
        func.__beets_config = config
        func.__beets_config_before_setup = before_setup

        @wraps(func)
        def apply_config(*args, **kwargs):
            if not func.__beets_config_before_setup:
                beets.config.add(config)
            func(*args, **kwargs)

        return apply_config
```
Additions to `setUp`

```python
def setUp(self):
    ...
    func = self._testMethodName
    if hasattr(func, '__beets_config') and func.__beets_config_before_setup:
        beets.config.add(func.__beets_config)
    ...
```

Using it
```python
class UseThePlugin(_common.TestCase):
    def setUp(self):
        super(UseThePlugin, self).setUp()
        # The config is now updated!
        self.plugin = fetchart.FetchArtPlugin()        
```

```python
class MyTest(UseThePlugin):
    @with_config({u'fetchart': {u'somekey': 42}})
    test_it(self)
        pass
```

If the undocumented attribute is not a showstopper, would this be worth including in `_common.TestCase`?

On a side-note; is there anyone working on https://github.com/beetbox/beets/issues/735#issuecomment-153418759 ? If not, I might take over joining `test._common`/`test.helper`.

**Comment by sampsyo** [comment]
Huh! Nifty idea. I definitely feel your pain---I've also run into cases where test cases share *most* of their `setUp` logic but want to use different configurations or other details. My solution is usually to invent my own setup function and calling it in each test case:

```py
def setUp(self):
    # only generic setup here

def my_set_up(self, config):
    beets.config.add(config)
    self.plugin = fetchart.FetchArtPlugin()

def test_a(self):
    self.my_set_up({u'fetchart': {u'somekey': 42}})
    # the actual test goes here
```

which is obviously less pretty than a decorator. But I'm slightly worried about the "magic" in the decorator approach: you usually expect `setUp` to do exactly the same thing for all tests. That might not be a big deal, though, especially compared to how frequently we set up temporary configurations.

And no, I don't think anyone's working on the unification of the test harness modules! That would be *awesome* if you're interested in looking into it.

**Comment by sampsyo** [comment]
Yes, sounds great. It's sort of a problem in general that we require `super(...).tearDown()` and such everywhere; it's not very pretty and it's error prone. But that's the way it is, at least for now.
